server_routers.go

```diff
@@ -0,0 +1,383 @@
+
+type composableRouter struct {
+	providers providersRouter
+	peers     peersRouter
+	ipns      ipnsRouter
+}
+
+func (r composableRouter) FindProviders(ctx context.Context, key cid.Cid, limit int) (iter.ResultIter[types.Record], error) {
+	if r.providers == nil {
+		return iter.ToResultIter(iter.FromSlice([]types.Record{})), nil
+	}
+	return r.providers.FindProviders(ctx, key, limit)
+}
+
+func (r composableRouter) FindPeers(ctx context.Context, pid peer.ID, limit int) (iter.ResultIter[*types.PeerRecord], error) {
+	if r.peers == nil {
+		return iter.ToResultIter(iter.FromSlice([]*types.PeerRecord{})), nil
+	}
+	return r.peers.FindPeers(ctx, pid, limit)
+}
+
+func (r composableRouter) GetIPNS(ctx context.Context, name ipns.Name) (*ipns.Record, error) {
+	if r.ipns == nil {
+		return nil, routing.ErrNotFound
+	}
+	return r.ipns.GetIPNS(ctx, name)
+}
+
+func (r composableRouter) PutIPNS(ctx context.Context, name ipns.Name, record *ipns.Record) error {
+	if r.ipns == nil {
+		return nil
+	}
+	return r.ipns.PutIPNS(ctx, name, record)
+}
+
+//lint:ignore SA1019 // ignore staticcheck
+func (r composableRouter) ProvideBitswap(ctx context.Context, req *server.BitswapWriteProvideRequest) (time.Duration, error) {
+	return 0, routing.ErrNotSupported
+}
+
+var _ server.ContentRouter = parallelRouter{}
+
+type parallelRouter struct {
+	routers []router
+}
+
+func (r parallelRouter) FindProviders(ctx context.Context, key cid.Cid, limit int) (iter.ResultIter[types.Record], error) {
+	return find(ctx, r.routers, func(ri router) (iter.ResultIter[types.Record], error) {
+		return ri.FindProviders(ctx, key, limit)
+	})
+}
+
+func (r parallelRouter) FindPeers(ctx context.Context, pid peer.ID, limit int) (iter.ResultIter[*types.PeerRecord], error) {
+	return find(ctx, r.routers, func(ri router) (iter.ResultIter[*types.PeerRecord], error) {
+		return ri.FindPeers(ctx, pid, limit)
+	})
+}
+
+func find[T any](ctx context.Context, routers []router, call func(router) (iter.ResultIter[T], error)) (iter.ResultIter[T], error) {
+	switch len(routers) {
+	case 0:
+		return iter.ToResultIter(iter.FromSlice([]T{})), nil
+	case 1:
+		return call(routers[0])
+	}
+
+	its := make([]iter.ResultIter[T], 0, len(routers))
+	var err error
+	for _, ri := range routers {
+		it, itErr := call(ri)
+
+		if itErr != nil {
+			err = errors.Join(err, itErr)
+		} else {
+			its = append(its, it)
+		}
+	}
+
+	// If all iterators failed to be created, then return the error.
+	if len(its) == 0 {
+		return nil, err
+	}
+
+	// Otherwise return manyIter with remaining iterators.
+	return newManyIter(ctx, its), nil
+}
+
+type manyIter[T any] struct {
+	ctx    context.Context
+	its    []iter.ResultIter[T]
+	nextCh chan int
+	next   int
+}
+
+func newManyIter[T any](ctx context.Context, its []iter.ResultIter[T]) *manyIter[T] {
+	nextCh := make(chan int)
+
+	for i, it := range its {
+		go func(ch chan int, it iter.ResultIter[T], index int) {
+			for it.Next() {
+				ch <- index
```

Comment:
You probably want a `select` that handles `ctx.Done` as well so that if the user doesn't call `Next` but the context cancels you don't get abandoned goroutines.